### PR TITLE
Add Docker scripts for automatic provisioning of Jenkins instance that is ready for testing

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

Add some Docker scripts that help to quickly spin up a local Jenkins instance that is ready for testing the plugin.

Some improvements compared to the previous dockerized version:
- admin setup step is skipped, all the necessary plugin dependencies are provisioned automatically
- agent nodes with various envs (.NET, Java, JS, Python) are created automatically
- Github credentials are created automatically (provided that an SSH key is supplied via an env var)
- some sample jobs and pipelines are provisioned automatically
- logs shipping is pre-configured in Datadog agent
- running `mvn clean package` no longer messes up the set of installed plugins, but only replaces the Datadog plugin
- a side effect of the previous point is that containers restart is significantly faster
- Jenkins master node exposes port 5055 for remote debugging

Development instructions are updated accordingly.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -26,108 +26,37 @@ At this point you're waiting on us. We may suggest some changes, improvements or
...
-1. Access your Jenkins instance http://localhost:8080
-1. Enter the administrator password in the Getting Started form.
-1. On the next page, click on the "Select plugins to be installed" unless you want to install all suggested plugins.
-1. Select desired plugins depending on your needs. You can always add plugins later.
-1. Create a user so that you don't have to use the admin credential again (optional).
-1. Continue until the end of the setup process and log back in.
-1. Go to http://localhost:8080/configure to configure the "Datadog Plugin", set your `API Key`.
-  - Click on the "Test Key" to make sure your key is valid.
+1. Optionally set the `GITHUB_SSH_KEY` and `GITHUB_SSH_KEY_PASSPHRASE` environment variables with the key and passphrase that can be used to access GitHub. This allows to automatically create GitHub credentials in Jenkins.   
+1. Run `mvn clean package` and `docker-compose -p datadog-jenkins-plugin -f docker/docker-compose.yaml up` from the directory where this repository is cloned/forked.
```
Comment (drodriguezhdez) [inline DEVELOPMENT.md:40]:
```suggestion
1. Run `mvn clean package -DskipTests` and `docker-compose -p datadog-jenkins-plugin -f docker/docker-compose.yaml up` from the directory where this repository is cloned/forked.
```

Maybe? In this case you just want to run the dev environment.

On the other hand, I've executed the command `mvn clean package` but it gets stuck in the `InjectionTest`

```
[INFO]
[INFO] --- maven-surefire-plugin:3.0.0-M7:test (default-test) @ datadog ---
[INFO] Using auto detected provider org.apache.maven.surefire.junit4.JUnit4Provider
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running InjectedTest
```

File: docker/docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
+version: '3.8'
+
+# The services below should only be used for development and testing purposes. They are not production-ready.
+services:
+  jenkins-master:
+    build:
+      context: ./docker/master-node
```
Comment (drodriguezhdez) [inline docker/docker-compose.yaml:7]:
```suggestion
      context: ./master-node
```

I think the paths need to be relative to the `docker-compose.yaml` file.

File: docker/docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      DATADOG_JENKINS_PLUGIN_TARGET_HOST: datadog # `dogstatsd` or `datadog` based on the container you wish to use
+      DATADOG_JENKINS_PLUGIN_TARGET_LOG_COLLECTION_PORT: 10518
+      DATADOG_JENKINS_PLUGIN_TARGET_API_KEY: $JENKINS_PLUGIN_DATADOG_API_KEY
+    volumes:
+      - jenkins_shared:/var/jenkins_home/shared
+      - $JENKINS_PLUGIN/target/datadog.hpi:/var/jenkins_home/plugins/datadog.hpi
+
+  jenkins-agent-dotnet:
+    build:
+      context: ./docker/agent-node
```
Comment (drodriguezhdez) [inline docker/docker-compose.yaml:28]:
```suggestion
      context: ./agent-node
```
Same as previous comment. Paths need to be relative to the `docker-compose.yaml`

File: docker/docker-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    environment:
+      JENKINS_URL: http://jenkins-master:8080
+      JENKINS_AGENT_NAME: agent-dotnet
+      JENKINS_AGENT_WORKDIR: /jenkins/agent
+    volumes:
+      - jenkins_shared:/shared
+
+  jenkins-agent-python:
+    build:
+      context: ./docker/agent-node
```
Comment (drodriguezhdez) [inline docker/docker-compose.yaml:41]:
```suggestion
      context: ./agent-node
```

File: docker/docker-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    environment:
+      JENKINS_URL: http://jenkins-master:8080
+      JENKINS_AGENT_NAME: agent-python
+      JENKINS_AGENT_WORKDIR: /jenkins/agent
+    volumes:
+      - jenkins_shared:/shared
+
+  jenkins-agent-js:
+    build:
+      context: ./docker/agent-node
```
Comment (drodriguezhdez) [inline docker/docker-compose.yaml:54]:
```suggestion
      context: ./agent-node
```

**Review by drodriguezhdez — COMMENTED** [review]
Huge improvement for plugin devs here! 

I think there are some errors on the file paths that point to the different `Dockerfile` files in the `docker-compose` file.

Also, let's honour the effort that Jenkins maintainers put on [deprecating the word `master`](https://community.jenkins.io/t/jenkins-2-319-1-master-controller-built-in-node-built-in/939) in favour of `controller` or `built-in` words for describing the nodes. 
I've left some suggestion comments for that, although there are more to change.

File: DEVELOPMENT.md (hunk 1 above)
Comment (nikita-tkachenko-datadog) [reply DEVELOPMENT.md:40]: Yeah, I don't think it actually gets stuck (for me it takes a few minutes to complete that test), but certainly running the tests is not a prerequisite for spinning up a local Jenkins instance.

File: docker/agent-node/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+JNLP_SECRET_PATH="/shared/$JENKINS_AGENT_NAME.secret"
+
+wait_for_jenkins_master() {
+    echo "Waiting for Jenkins master at ${JENKINS_URL} to be available..."
```
Comment (nikita-tkachenko-datadog) [inline docker/agent-node/entrypoint.sh:6]:
```suggestion
    echo "Waiting for Jenkins controller at ${JENKINS_URL} to be available..."
```

File: docker/agent-node/entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    while true; do
+        if [ -f "$JNLP_SECRET_PATH" ]; then
+            break
+        else
+            printf '.'
+            sleep 2
+        fi
+    done
+
+    echo "Jenkins master is up."
```
Comment (nikita-tkachenko-datadog) [inline docker/agent-node/entrypoint.sh:22]:
```suggestion
    echo "Jenkins controller is up."
```

File: docker/agent-node/entrypoint.sh [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+wait_for_jenkins_master
```
Comment (nikita-tkachenko-datadog) [inline docker/agent-node/entrypoint.sh:25]:
```suggestion
wait_for_jenkins_controller
```

File: docker/docker-compose.yaml (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply docker/docker-compose.yaml:7]: It's relative to the build context, which by default is the current folder (the one where the build command is executed).

The command given in the docs is expected to be run from the repo root:
> run `docker-compose -p datadog-jenkins-plugin -f docker/docker-compose.yaml up` from **the directory where this repository is cloned**

File: docker/docker-compose.yaml [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    volumes:
+      - jenkins_shared:/var/jenkins_home/shared
+      - $JENKINS_PLUGIN/target/datadog.hpi:/var/jenkins_home/plugins/datadog.hpi
+
+  jenkins-agent-dotnet:
+    build:
+      context: ./docker/agent-node
+      dockerfile: Dockerfile-dotnet
+    depends_on:
+      - jenkins-master
```
Comment (nikita-tkachenko-datadog) [inline docker/docker-compose.yaml:31]:
```suggestion
      - jenkins-controller
```

File: docker/docker-compose.yaml [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    environment:
+      JENKINS_URL: http://jenkins-master:8080
```
Comment (nikita-tkachenko-datadog) [inline docker/docker-compose.yaml:33]:
```suggestion
      JENKINS_URL: http://jenkins-controller:8080
```

File: docker/docker-compose.yaml [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+      JENKINS_AGENT_WORKDIR: /jenkins/agent
+    volumes:
+      - jenkins_shared:/shared
+
+  jenkins-agent-python:
+    build:
+      context: ./docker/agent-node
+      dockerfile: Dockerfile-python
+    depends_on:
+      - jenkins-master
```
Comment (nikita-tkachenko-datadog) [inline docker/docker-compose.yaml:44]:
```suggestion
      - jenkins-controller
```

File: docker/docker-compose.yaml [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    environment:
+      JENKINS_URL: http://jenkins-master:8080
```
Comment (nikita-tkachenko-datadog) [inline docker/docker-compose.yaml:46]:
```suggestion
      JENKINS_URL: http://jenkins-controller:8080
```

File: docker/docker-compose.yaml [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+      JENKINS_AGENT_WORKDIR: /jenkins/agent
+    volumes:
+      - jenkins_shared:/shared
+
+  jenkins-agent-js:
+    build:
+      context: ./docker/agent-node
+      dockerfile: Dockerfile-js
+    depends_on:
+      - jenkins-master
```
Comment (nikita-tkachenko-datadog) [inline docker/docker-compose.yaml:57]:
```suggestion
      - jenkins-controller
```

File: docker/docker-compose.yaml [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    environment:
+      JENKINS_URL: http://jenkins-master:8080
```
Comment (nikita-tkachenko-datadog) [inline docker/docker-compose.yaml:59]:
```suggestion
      JENKINS_URL: http://jenkins-controller:8080
```

File: docker/docker-compose.yaml (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply docker/docker-compose.yaml]: Indeed it fails with the newer versions of `docker-compose`. Fixed the paths and added a note to the doc saying that `docker-compose` needs to be updated if there are "path ... not found" errors.

**Review by drodriguezhdez — APPROVED** [review]

**Pull request merged**